keps/sig-scheduling/2044-topology-awareness-in-kube-scheduler/README.md

```diff
@@ -0,0 +1,416 @@
+[Topology Manager](https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/) was
+introduced in kubernetes as part of kubelet in order to extract the best performance out of
+these high performance hybrid systems. It performs optimizations related to resource allocation
+in order to make it more likely for a given pod to perform optimally. In scenarios where
+Topology Manager is unable to align topology of requested resources based on the selected
+Topology Manager policy, the pod is rejected with Topology Affinity Error.
+[This](https://github.com/kubernetes/kubernetes/issues/84869) kubernetes issue provides
+further context on how runaway pods are created because the scheduler is topology-unaware.
+
+In order to address this issue, scheduler needs to choose a node considering resource availability
+along with underlying resource topology and Topology Manager policy on the worker node.
+
+This document describes behaviour of the Kubernetes Scheduler which takes worker node topology into account.
+
+# Motivation
+
+After Topology Manager was introduced, the problem of launching pod in the cluster where worker
+nodes have different NUMA topology and different amount of resources in that topology became
+actual. Pod could be scheduled on the node where total amount of resources are enough, but
+resource distribution could not satisfy the appropriate Topology policy. In this case the pod
+failed to start. Much better behaviour for scheduler would be to select appropriate node where
+kubelet admit handlers may pass.
+
+In order to enable topology aware scheduling in Kubernetes, resource topology information of the
+nodes in the cluster needs to be exposed to the scheduler so that it can use it to make a more
+informed scheduling decision. This KEP describes how it would be implemented.
+
+## Goals
+
+- Make scheduling process more precise when we have NUMA topology on the
+worker node.
+- Enhance the node object to capture topology information which can be referred to 
+by the scheduler.
+
+## Non-Goals
+
+- Change the PodSpec to allow requesting a specific node topology manager policy
+- This Proposal requires exposing NUMA topology information. This KEP doesn't
+describe how to expose all necessary information it just declare what kind of
+information is necessary.
+- Changes to the TopologyManager and its policies.
+
+# Proposal
+
+Kube-scheduler plugin will be moved from kuberntes-sigs/scheduler-plugin (or out-of-tree)
+plugin into the main tree as a built-in plugin. This plugin implements a simplified version of Topology Manager and hence is different from original topology manager algorithm. Plugin would
+be disabled by default and when enabled would check for the ability to run pod only in case of single-numa-node policy on the node, since it is the most strict policy, it implies that the launch on the node with other existing policies will be successful if the condition for single-numa-node policy passed for the worker node.
```

Comment:
It is a sys-admin decision to configure kubelet with Topology Manager policy. Not all workloads require strict NUMA alignment and therefore Topology Manager policy could be configured to be none or restricted. More information on this is [here](https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/#topology-manager-policies). The default policy is `none` where Topology Manager does not perform any topology alignment. But for a workload that requires NUMA alignment a node where `single-numa-node policy` is not configured would lead to underperformance of the workload.